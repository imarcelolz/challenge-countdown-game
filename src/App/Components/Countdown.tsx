import { CountdownGame } from '@/Lib/CountdownGame/CountdownGame';
import { CountdownGameState } from '@/Lib/CountdownGame/CountdownGame.types';
import React, { Component } from 'react';
import { CountdownState } from './Countdown.types';
import { Clock } from './Views/Clock';
import { WordPanel } from './Views/WordPanel';

const InitialState = (): CountdownState => ({
  clockSeconds: 0,
  randomWord: [],
  wordLength: 0,
  answers: [],
});

export class Countdown extends Component<{}, CountdownState> {
  private game: CountdownGame;
  private set ClockSeconds(clockSeconds: number) { this.setState({ clockSeconds }); }
  private get ClockSeconds(): number { return this.state.clockSeconds; }

  constructor(props) {
    super(props);

    this.game = new CountdownGame(this.onGameOver, this.onClockTick);
    this.state = InitialState();
  }

  onStartClick = () => {
    const gameState = this.game.start();

    const newState = InitialState();
    newState.randomWord = gameState.randomWord.split('')
      .map((item) => ({ value: item }));

    newState.wordLength = gameState.wordLength;

    this.setState(newState);
  };

  onStopClick = () => {
    this.ClockSeconds = 0;
    this.game.stop();
  };

  onGameOver = (gameState: CountdownGameState) => {
    let message;

    switch (gameState.result) {
      case 'success':
        message = 'Congratulations, you win!';
        break;
      case 'timeout':
        message = 'Game over, you will need to be hurry on the next time';
        break;
      case 'too-many-tries':
        message = 'Game over, you reach the error limit, try again';
        break;
    }

    this.setState({
      answers: gameState.answers,
      matchResult: gameState.result,
      message
    });
  };

  onClockTick = () => {
    this.ClockSeconds += 1;
  }

  onOriginSelected = (index: number) => {
    const { randomWord } = this.state;
    const updated = [...randomWord];

    const item = updated[index];
    if(item.selected) {
      return;
    }

    item.selected = true;

    this.setState({ selected: item, randomWord });
  }

  onDestinationSelected = (index: number) => {
    const { selected } = this.state;
    if(!selected) { return; }

    const answers = this.game.addLetter({ value: selected.value, index });

    this.setState({ answers, selected });
  }

  private get CurrentAnswer() {
    const { wordLength, answers } = this.state;

    const currentAnswer = [];
    for(let i = 0 ; i < wordLength; i++) {
      const answer = answers?.find((item) => item.index === i);

      currentAnswer.push(answer || { value: '' });
    }

    return currentAnswer;
  }

  render() {
    const { randomWord, message } = this.state;

    return (
      <div className="countdown-container">
        <Clock seconds={this.ClockSeconds} />
        <span>
          <h3>Create the word by dragging letters into empty boxes</h3>
          <h5>You have one minute</h5>
        </span>
        <WordPanel className="origin" letters={randomWord} dragDirection="from" onSelect={this.onOriginSelected} />
        <WordPanel className="destination" letters={this.CurrentAnswer} dragDirection="to" onSelect={this.onDestinationSelected}/>

        <h5>{message}</h5>
        <div>
          <button className='start' onClick={this.onStartClick}>Start</button>
          <button className='stop' onClick={this.onStopClick}>Stop</button>
        </div>
      </div>
    );
  }
}
