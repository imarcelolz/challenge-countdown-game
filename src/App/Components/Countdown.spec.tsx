import { mount, shallow } from 'enzyme';
import React from 'react';
import { Countdown } from './Countdown';

let wrapper;
describe('Context: Rendering', () => {
  it('renders without crashing', () => {
    expect(shallow(<Countdown />)).toMatchSnapshot();
  });
});

describe('Context: Functional', () => {
  let game;

  beforeEach(() => {
    wrapper = mount(<Countdown />);
    game = wrapper.instance().game;
  });

  it('starts the game on start Click', () => {
    const spy = jest.spyOn(game, 'start');
    wrapper.find('button.start').invoke('onClick')();

    expect(spy).toHaveBeenCalledTimes(1);
  });

  xit('stops the game on stop click', () => {
    const spy = jest.spyOn(game, 'stop');
    wrapper.find('button.stop').invoke('onClick')();

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('updates the destination on selection', () => {
    const spy = jest.spyOn(game, 'addLetter');

    wrapper.find('button.start').invoke('onClick')();
    wrapper.find('.origin').invoke('onSelect')(0);
    wrapper.find('.destination').invoke('onSelect')(0)

    expect(spy).toHaveBeenCalled();
  });
});
