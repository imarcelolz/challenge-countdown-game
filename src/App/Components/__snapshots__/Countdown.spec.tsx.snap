// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context: Rendering renders without crashing 1`] = `ShallowWrapper {}`;
